### ISSUE TITLE
Significant performance degradation post-2.1.2 in Lucene query

### ISSUE THREAD

**Issue by cory-p-oncota** [orientechnologies/orientdb#5710]
In one of our prototyping projects utilizing OrientDB, we've noticed a significant performance degradation since moving off of version 2.1.2. In my memory, our pathway was 2.1.5, 2.1.8, and 2.1.10. The primary hit is against a Lucene index covering approximately 30 million vertices.

The specific query in question is as follows:

```
select from ( select *, expand(in("anatomy_map")) from umls_concept where description lucene " [SEARCH_TERM]" limit 100) TIMEOUT 5000
```

On an AWS EC2 r3.xlarge, processing a document in our service has grown from about 15 seconds to 5 minutes, with numerous timeouts occurring. All servers were running the same configuration. Any advice on what may have caused this regression or a variation in configuration which may solve the problem?

**Comment by robfrank** [comment]
In order to understand better I need more details:
- db's schema (relevant part)
- sample data (anonymized and/or generated)
- sample queries (with (almost) real use cases)
- an 'ls -la' of the lucene dir under the database dir: useful to understand if the index is merged or not and the size of the index
- orientDb deployment: standalone or embedded?
- an htop screenshot to understand ram usage
- Is the db growing from time to time? 
thanks in advance

**Comment by robfrank** [comment]
In order to understand better I need more details:
- db's schema (relevant part)
- sample data (anonymized and/or generated)
- sample queries (with (almost) real use cases)
- an 'ls -la' of the lucene dir under the database dir: useful to understand if the index is merged or not and the size of the index
- orientDb deployment: standalone or embedded?
- an htop screenshot to understand ram usage
- Is the db growing from time to time? 
thanks in advance

**Comment by cory-p-oncota** [comment]
The relevant schema is as such:
```
VERTEX umls_concept
  termType LINK
  lui STRING
  source LINK
  description STRING
  isPreferredLui BOOLEAN
  cui STRING
  code STRING
  isPreferredTerm BOOLEAN
  aui STRING

EDGE anatomy_map (lightweight)

VERTEX anatomy_key
  value_type STRING
  path STRING
  display STRING
```

A umls_concept sample record (1 of 7.6 million):
```json
{
    "result": [
        {
            "@type": "d",
            "@rid": "#149:127525",
            "@version": 31,
            "@class": "clinical_attribute",
            "termType": "#992:184",
            "lui": "L2943147",
            "source": "#36:96",
            "description": "pT category",
            "isPreferredLui": "false",
            "cui": "C1270004",
            "code": "384625004",
            "isPreferredTerm": "true",
            "aui": "A3804415",
            "in_anatomy_map": [
                "#1023:2772"
            ],
            "out_isa": [
                "#988:780488",
                "#988:1220752"
            ],
            "in_isa": [
                "#988:1237732",
                "#988:1237733",
                "#988:1237734",
                "#988:1237735",
                "#988:1237736",
                "#988:1237737",
                "#988:1237738",
                "#988:1237739",
                "#988:1237740",
                "#988:1237741",
                "#988:1237742",
                "#988:1237743",
                "#988:1237744"
            ],
            "@fieldTypes": "termType=x,source=x,in_anatomy_map=g,out_isa=g,in_isa=g"
        }
    ],
    "notification": "Query executed in 0.044 sec. Returned 1 record(s)"
}
```

And a sample anatomy_key record (1 of ~100):
```json
{
    "result": [
        {
            "@type": "d",
            "@rid": "#1021:852",
            "@version": 2,
            "@class": "anatomy_key",
            "value_type": "option",
            "path": "t",
            "display": "t",
            "out_anatomy_map": [
                "#1023:2772"
            ],
            "@fieldTypes": "out_anatomy_map=g"
        }
    ],
    "notification": "Query executed in 0.042 sec. Returned 1 record(s)"
}
```

And the index coverage on the class is as follows where full_text_search is the Lucene index involved:
```
-------------------------------+----------------+
 NAME                          | PROPERTIES     |
-------------------------------+----------------+
 full_text_search              | description    |
 code_lookup                   | source (+)     |
                               | code           |
 umls_concept_unique_constraint| aui            |
 umls_source_lookup            | source         |
 concept_id                    | cui            |
 term_filter                   | termType (+)   |
                               | description    |
-------------------------------+----------------+
```

As far as queries go, the one listed in the original issue above is the only one executed against this subgraph. Queries running against another subgraph which do not use Lucene indexes are performing equally between versions.

Here's the result of `ls -la`:
```
total 213056
drwxrwxr-x 2 ubuntu ubuntu     4096 Feb 10 23:17 .
drwxrwxr-x 5 ubuntu ubuntu     4096 Feb 10 21:19 ..
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:21 _10.cfe
-rw-rw-r-- 1 ubuntu ubuntu  2301456 Feb 10 21:21 _10.cfs
-rw-rw-r-- 1 ubuntu ubuntu      269 Feb 10 21:21 _10.si
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:21 _11.cfe
-rw-rw-r-- 1 ubuntu ubuntu  5727993 Feb 10 21:21 _11.cfs
-rw-rw-r-- 1 ubuntu ubuntu      269 Feb 10 21:21 _11.si
-rw-rw-r-- 1 ubuntu ubuntu  6770925 Feb 10 21:21 _12.fdt
-rw-rw-r-- 1 ubuntu ubuntu    18394 Feb 10 21:21 _12.fdx
-rw-rw-r-- 1 ubuntu ubuntu      216 Feb 10 21:21 _12.fnm
-rw-rw-r-- 1 ubuntu ubuntu 10592779 Feb 10 21:21 _12_Lucene41_0.doc
-rw-rw-r-- 1 ubuntu ubuntu  6058349 Feb 10 21:21 _12_Lucene41_0.pos
-rw-rw-r-- 1 ubuntu ubuntu 20783257 Feb 10 21:21 _12_Lucene41_0.tim
-rw-rw-r-- 1 ubuntu ubuntu   320135 Feb 10 21:21 _12_Lucene41_0.tip
-rw-rw-r-- 1 ubuntu ubuntu  1586864 Feb 10 21:21 _12.nvd
-rw-rw-r-- 1 ubuntu ubuntu       46 Feb 10 21:21 _12.nvm
-rw-rw-r-- 1 ubuntu ubuntu      407 Feb 10 21:21 _12.si
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:21 _13.cfe
-rw-rw-r-- 1 ubuntu ubuntu  5816102 Feb 10 21:21 _13.cfs
-rw-rw-r-- 1 ubuntu ubuntu      269 Feb 10 21:21 _13.si
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:22 _1d.cfe
-rw-rw-r-- 1 ubuntu ubuntu  4661516 Feb 10 21:22 _1d.cfs
-rw-rw-r-- 1 ubuntu ubuntu      269 Feb 10 21:22 _1d.si
-rw-rw-r-- 1 ubuntu ubuntu  7267857 Feb 10 21:22 _1e.fdt
-rw-rw-r-- 1 ubuntu ubuntu    19442 Feb 10 21:22 _1e.fdx
-rw-rw-r-- 1 ubuntu ubuntu      216 Feb 10 21:22 _1e.fnm
-rw-rw-r-- 1 ubuntu ubuntu 10428766 Feb 10 21:22 _1e_Lucene41_0.doc
-rw-rw-r-- 1 ubuntu ubuntu  5582178 Feb 10 21:22 _1e_Lucene41_0.pos
-rw-rw-r-- 1 ubuntu ubuntu 20714424 Feb 10 21:22 _1e_Lucene41_0.tim
-rw-rw-r-- 1 ubuntu ubuntu   312888 Feb 10 21:22 _1e_Lucene41_0.tip
-rw-rw-r-- 1 ubuntu ubuntu  1704553 Feb 10 21:22 _1e.nvd
-rw-rw-r-- 1 ubuntu ubuntu       46 Feb 10 21:22 _1e.nvm
-rw-rw-r-- 1 ubuntu ubuntu      407 Feb 10 21:22 _1e.si
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:22 _1f.cfe
-rw-rw-r-- 1 ubuntu ubuntu  4316474 Feb 10 21:22 _1f.cfs
-rw-rw-r-- 1 ubuntu ubuntu      269 Feb 10 21:22 _1f.si
-rw-rw-r-- 1 ubuntu ubuntu  6525747 Feb 10 21:20 _c.fdt
-rw-rw-r-- 1 ubuntu ubuntu    17367 Feb 10 21:20 _c.fdx
-rw-rw-r-- 1 ubuntu ubuntu      216 Feb 10 21:20 _c.fnm
-rw-rw-r-- 1 ubuntu ubuntu  8163323 Feb 10 21:20 _c_Lucene41_0.doc
-rw-rw-r-- 1 ubuntu ubuntu  5185129 Feb 10 21:20 _c_Lucene41_0.pos
-rw-rw-r-- 1 ubuntu ubuntu 21252254 Feb 10 21:20 _c_Lucene41_0.tim
-rw-rw-r-- 1 ubuntu ubuntu   336414 Feb 10 21:20 _c_Lucene41_0.tip
-rw-rw-r-- 1 ubuntu ubuntu  1529745 Feb 10 21:20 _c.nvd
-rw-rw-r-- 1 ubuntu ubuntu       46 Feb 10 21:20 _c.nvm
-rw-rw-r-- 1 ubuntu ubuntu      397 Feb 10 21:20 _c.si
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:20 _h.cfe
-rw-rw-r-- 1 ubuntu ubuntu   848553 Feb 10 21:20 _h.cfs
-rw-rw-r-- 1 ubuntu ubuntu      266 Feb 10 21:20 _h.si
-rw-rw-r-- 1 ubuntu ubuntu  7524873 Feb 10 21:20 _p.fdt
-rw-rw-r-- 1 ubuntu ubuntu    20476 Feb 10 21:20 _p.fdx
-rw-rw-r-- 1 ubuntu ubuntu      216 Feb 10 21:20 _p.fnm
-rw-rw-r-- 1 ubuntu ubuntu 13407327 Feb 10 21:20 _p_Lucene41_0.doc
-rw-rw-r-- 1 ubuntu ubuntu  7119002 Feb 10 21:20 _p_Lucene41_0.pos
-rw-rw-r-- 1 ubuntu ubuntu 19770661 Feb 10 21:20 _p_Lucene41_0.tim
-rw-rw-r-- 1 ubuntu ubuntu   283177 Feb 10 21:20 _p_Lucene41_0.tip
-rw-rw-r-- 1 ubuntu ubuntu  1763732 Feb 10 21:20 _p.nvd
-rw-rw-r-- 1 ubuntu ubuntu       46 Feb 10 21:20 _p.nvm
-rw-rw-r-- 1 ubuntu ubuntu      397 Feb 10 21:20 _p.si
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:21 _r.cfe
-rw-rw-r-- 1 ubuntu ubuntu  5290713 Feb 10 21:21 _r.cfs
-rw-rw-r-- 1 ubuntu ubuntu      266 Feb 10 21:21 _r.si
-rw-rw-r-- 1 ubuntu ubuntu      484 Feb 10 21:22 segments_2
-rw-rw-r-- 1 ubuntu ubuntu       20 Feb 10 21:22 segments.gen
-rw-rw-r-- 1 ubuntu ubuntu        0 Feb 10 23:17 write.lock
-rw-rw-r-- 1 ubuntu ubuntu      268 Feb 10 21:21 _x.cfe
-rw-rw-r-- 1 ubuntu ubuntu  3931374 Feb 10 21:21 _x.cfs
-rw-rw-r-- 1 ubuntu ubuntu      266 Feb 10 21:21 _x.si
```

We're running a standalone remote instance, currently non-distributed.

A capture of htop (http://postimg.org/image/wtt4k3dwt/). Note, this isn't under load at the moment, however, I've rarely seen usage exceed 13%. I am noticing now that I'm actually allocating less heap and disk cache when using 2.1.2. which doesn't seem like a root cause but just in case the settings are:
```
2.1.2 - 512MB max heap and 1.5GB disk cache
2.1.2+ - 4GB max heap and 8GB disk cache
```

Finally, this database is actually entirely read-only with about 11GB of data, so perhaps some more ideal memory settings exist for that scenario.

**Comment by maggiolo00** [comment]
hi @cory-p-oncota 

can you post the explain of this query?

```
select from ( select *, expand(in("anatomy_map")) from umls_concept where description lucene " [SEARCH_TERM]" limit 100) TIMEOUT 5000
```

**Comment by cory-p-oncota** [comment]
Under 2.1.2:
```json
{
    "result": [
        {
            "@type": "d",
            "@version": 0,
            "score": 2.9743857,
            "current": "#145:194294",
            "totalHits": 140,
            "limit": 100,
            "luceneIndex": true,
            "fetchingFromTargetElapsed": 0,
            "expandElapsed": 2,
            "full_text_search_totalHits": 140,
            "user": "#5:0",
            "elapsed": 7.779243,
            "resultType": "collection",
            "resultSize": 0,
            "@fieldTypes": "score=f,current=x,fetchingFromTargetElapsed=l,expandElapsed=l,user=x,elapsed=f"
        }
    ],
    "notification": "Query executed in 0.026 sec. Returned 1 record(s)"
}
```

And under 2.1.10:
```json
{
    "result": [
        {
            "@type": "d",
            "@version": 0,
            "score": 1.4871929,
            "current": "#145:194098",
            "totalHits": 140,
            "_memoryIndex": "description:\n\t'[32]':2: [(5),(24)]\n\t'[35]':1: [(11)]\n\t'[36]':1: [(31)]\n\t'[63 61 72 63 69 6e 6f 6d 61]':1: [(35)]\n\t'[63 6d]':2: [(6),(12)]\n\t'[64 65]':6: [(2),(4),(10),(30),(33),(36)]\n\t'[64 65 72 6d 69 73]':1: [(20)]\n\t'[64 69 6d 65 6e 73 69 c3 b3 6e]':1: [(16)]\n\t'[65 6e]':2: [(13),(26)]\n\t'[65 73 70 65 73 6f 72]':2: [(27),(34)]\n\t'[65 78 63 6c 75 69 64 6f]':1: [(38)]\n\t'[68 61 6c 6c 61 7a 67 6f]':1: [(44)]\n\t'[6c 61]':1: [(19)]\n\t'[6c 69 6d 69 74 61 64 6f]':1: [(17)]\n\t'[6d 61 79 6f 72]':3: [(9),(15),(22)]\n\t'[6d 65 6c 61 6e 6f 6d 61]':1: [(43)]\n\t'[6d 6d]':2: [(25),(32)]\n\t'[6d c3 a1 73]':2: [(3),(29)]\n\t'[70 65 6e 65]':1: [(41)]\n\t'[70 65 72 6f]':2: [(7),(28)]\n\t'[70 69 65 6c]':1: [(37)]\n\t'[70 74 32 62]':1: [(0)]\n\t'[70 c3 a1 72 70 61 64 6f]':1: [(39)]\n\t'[73 75]':1: [(14)]\n\t'[74 75 6d 6f 72]':1: [(1)]\n\t'[76 75 6c 76 61]':1: [(40)]\n\t'[79]':2: [(21),(42)]\n\tterms=27, positions=42, memory=33.2 KB\n\nfields=1, terms=27, positions=42, memory=66 KB",
            "limit": 100,
            "luceneIndex": true,
            "fetchingFromTargetElapsed": 0,
            "expandElapsed": 13,
            "full_text_search_totalHits": 140,
            "user": "#5:0",
            "elapsed": 129.28113,
            "resultType": "collection",
            "resultSize": 0,
            "@fieldTypes": "score=f,current=x,fetchingFromTargetElapsed=l,expandElapsed=l,user=x,elapsed=f"
        }
    ],
    "notification": "Query executed in 0.158 sec. Returned 1 record(s)"
}
```

**Comment by robfrank** [comment]
Hi, thank you for your detailed report. We have identified the problem, I hope we could find a solution in a week.

**Comment by lvca** [comment]
@robfrank any news on this?

**Comment by robfrank** [comment]
I close this issue even if the "problem" still affect 2.2.x, but it is resolved on 3.0.x by the new SQL engine
